### PR TITLE
mrc-4629: Add function to pull a packet from a location

### PR DESCRIPTION
**Pull request by r-ash**
This PR will
* Add `outpack_location_pull_packet` to pull a packet from a specified location. Equivalent to `orderly_location_pull_packet` in R impl. And follows a very similar structure to that.

This is quite a big PR sorry about that, but includes around 300 lines which I have moved from 1 file to another. Let me know if still too much and I can try and split this up a little.

Still a few things TODO but this looks big enough as is
* Add support for running this with a query, and write the tests for that
* Add an interface for the location driver

Next up
* Pushing a packet to a location

### CODE REVIEW

File: src/outpack/filestore.py [hunk 1]
```diff
@@ -2,39 +2,44 @@
...
 
 class FileStore:
     def __init__(self, path):
-        self._path = path
+        self._path = Path(path)
         os.makedirs(path, exist_ok=True)
 
     def filename(self, hash):
         dat = hash_parse(hash)
-        return os.path.join(
```
Comment (richfitz) [inline src/outpack/filestore.py:16]: probably worth making a ticket to find remaining uses of os.path.join and replace them all

File: src/outpack/index.py [hunk 2]
```diff
@@ -62,7 +74,7 @@ def _read_metadata(path_root, data):
     return data
 
 
-def _read_locations(path_root, data):
+def _read_locations(path_root, data) -> Dict[str, Dict[str, PacketLocation]]:
```
Comment (richfitz) [inline src/outpack/index.py:77]: If we're going to go through and add type annotations everywhere we should probably do one big push in another PR; probably on the arguments too?

File: src/outpack/location.py [hunk 3]
```diff
@@ -250,3 +140,7 @@ def _location_driver(location_name, root):
     elif location.type == "custom":  # pragma: no cover
         msg = "custom remote not yet supported"
         raise Exception(msg)
+
+
+# TODO: Create a driver interface and have location path extend it?
```
Comment (richfitz) [inline src/outpack/location.py]: is there a ticket for this, perhaps with some more context as I don't follow enough to implement this myself if I need to (though it may be more obvious than I think?)

File: src/outpack/location_pull.py [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+                known_there[packet_id].time,
+            )
+
+
+def outpack_location_pull_packet(
+    ids: Union[str, List[str]],
+    options=None,
+    recursive=None,
+    root=None,
+    *,
```
Comment (richfitz) [inline src/outpack/location_pull.py]: should this star go earlier (after ids)?

File: src/outpack/location_pull.py [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+    print(
+        f"Unpacked {n_packets} {pl(n_packets, 'packet')} in "
+        f"{humanize.time.precisedelta(int(time.time() - time_start))}."
+    )
+
+    cleanup()
+    return list(plan.packets.keys())
+
+
+# This approach may be suboptimal in the case where the user does not
```
Comment (richfitz) [inline src/outpack/location_pull.py:174]: I recognise this comment :)

File: src/outpack/location_pull.py [hunk 6]
```diff
@@ -0,0 +1,463 @@
...
+def _location_pull_files(
+    files: List[PacketFileWithLocation], root: OutpackRoot
+) -> Tuple[FileStore, Callable[[], None]]:
+    store = root.files
+    if store is not None:
+
+        def cleanup():
+            return None
+
+        exists, missing = partition(lambda file: store.exists(file.hash), files)
```
Comment (richfitz) [inline src/outpack/location_pull.py:198]: yasss, lamdas.

Comment (richfitz) [inline src/outpack/location_pull.py:198]: Is `missing` unused? surprised you don't get lintr errors about that

```suggestion
        exists, _ = partition(lambda file: store.exists(file.hash), files)
```

File: src/outpack/location_pull.py [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+
+        if len(exists) > 0:
```
Comment (richfitz) [inline src/outpack/location_pull.py]:
```suggestion
        if exists:
```

I think this is the pythonic way of doing these tests

File: src/outpack/location_pull.py [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+
+
+def _location_pull_hash_store(
+    files: List[PacketFileWithLocation],
+    location_name: str,
+    driver,
+    store: FileStore,
+):
+    no_of_files = len(files)
+    # TODO: show a nice progress bar for users
```
Comment (richfitz) [inline src/outpack/location_pull.py:257]: yeah - in general it would be nice to find a suitable python `cli`-like packages for all the print statements and things like progress bars

File: src/outpack/location_pull.py [hunk 9]
```diff
@@ -0,0 +1,463 @@
...
+
+
+def _location_build_pull_plan_packets(
+    packet_ids: List[str], root: OutpackRoot, *, recursive: Optional[bool]
+) -> PullPlanPackets:
+    requested = packet_ids
+    if recursive is None:
+        recursive = root.config.core.require_complete_tree
+    if root.config.core.require_complete_tree and not recursive:
+        msg = (
```
Comment (richfitz) [inline src/outpack/location_pull.py]: Nicer with `"""` perhaps?

File: src/outpack/location_pull.py [hunk 10]
```diff
@@ -0,0 +1,463 @@
...
+        return []
+
+    # Find first location within the set which contains each packet
+    # We've already checked earlier that the file is in at least 1
+    # location so we don't have to worry about that here
+    all_files = []
+    seen_hashes = set()
+    for location_name in locations:
+        location_packets = set(root.index.packets_in_location(location_name))
+        packets_in_location = location_packets & packet_ids
```
Comment (richfitz) [inline src/outpack/location_pull.py]: Does this do set intersection? Did we decide on the more explicit `.intersection()` method before?

**Review by richfitz — CHANGES_REQUESTED** [review]
Thanks Rob - looks like this is a pretty direct translation from the R version? At least in places? That does seem to work at least.

A few comments here on things we might do in other tickets - notably to complete the work adding type annotations everywhere as we're all over the show on this atm and that's causing some churn.

Paul will likely have good comments on pythonic ways of doing things

File: src/outpack/filestore.py [hunk 11]
```diff
@@ -2,39 +2,44 @@
...
-            self._path, dat.algorithm, dat.value[:2], dat.value[2:]
-        )
+        return self._path / dat.algorithm / dat.value[:2] / dat.value[2:]
 
-    def get(self, hash, dst):
+    def get(self, hash, dst, overwrite):
```
Comment (plietar) [inline src/outpack/filestore.py]:
```suggestion
    def get(self, hash, dst, *, overwrite):
```

Generally nice for booleans to be kw arguments only, as it makes the call site explicit. Could give it a default too.

File: src/outpack/location_pull.py (hunk 6 above)
Comment (plietar) [reply src/outpack/location_pull.py:198]: If it isn't being used then this could be a `filter` instead right?

File: src/outpack/location_pull.py [hunk 12]
```diff
@@ -0,0 +1,463 @@
...
+        if len(exists) > 0:
+            print(
+                f"Found {len(exists)} {pl(exists, 'file')} in the "
+                f"file store"
+            )
+    else:
+        print("Looking for suitable files already on disk")
+        store = _temporary_filestore(root)
+
+        def cleanup():
```
Comment (plietar) [inline src/outpack/location_pull.py]: This is kind of surprising and I'm honestly surprised it even works.

```python
if ...:
  cleanup = lambda: None
else:
  cleanup = lambda: store.destroy()
```

this is a little more clear IMO, still not great.

I think this would be nicer structured as a contextmanager. It makes sure the caller doesn't forget to call the cleanup on all code paths, including on exceptions.

File: src/outpack/filestore.py [hunk 13]
```diff
@@ -43,9 +48,36 @@ def ls(self):
...
+                func(path)
+            else:
+                raise
+
+        shutil.rmtree(self._path, onerror=onerror)
+
+    def tmp(self) -> str:
+        path = self._path / "tmp"
+        path.mkdir(exist_ok=True)
+        return tempfile.NamedTemporaryFile(dir=path).name
```
Comment (plietar) [inline src/outpack/filestore.py]: This pattern is a bit dangerous. `NamedTemporaryFile` will delete the file when it gets garbage collected (or closed).
The GC could happen before the caller has started using this file (which is kind of okay I guess), long after the caller is done with it (which is fine but leaves a bunch of rubbish on the disk). The worst case is the GC run halfway through the caller using the path, and breaks everything by deleting the file.

The safe way of doing this is to return the `NamedTemporaryFile` object here, and at the call site use it within a `with` block:
```py
with store.tmp() as f:
  do_stuff(f.name)
```

If you don't want to expose the `NamedTemporaryFile` class you can also make your own context manager:
```py
from contextlib import contextmanager

@contextmanager
def tmp(self):
    path = self._path / "tmp"
    path.mkdir(exist_ok=True)
    with tempfile.NamedTemporaryFile(dir=path) as f:
        yield f.name

with store.tmp() as path:
    do_stuff(path)
```

Note the yield instead of return. The function will stop at this point, pass the value to the caller, then when the caller exists the with block the `tmp` function resumes. It itself leaves its `with` block, which deletes the file in a controlled and deterministic way.

(I don't remember off the top of my head how type annotations work on context managers)

File: src/outpack/location_pull.py (hunk 8 above)
Comment (plietar) [reply src/outpack/location_pull.py:257]: Came across https://click.palletsprojects.com/en/8.1.x/ the other day, looks promising.

File: src/outpack/location_pull.py [hunk 14]
```diff
@@ -0,0 +1,463 @@
...
+            f"dependencies of those requested"
+        )
+    if plan.info.n_skip > 0:
+        print(
+            f"Skipping {plan.info.n_skip} of {plan.info.n_total} "
+            f"{pl(plan.info.n_total, 'packet')} which are already "
+            f"unpacked"
+        )
+
+    store, cleanup = _location_pull_files(plan.files, root)
```
Comment (plietar) [inline src/outpack/location_pull.py]: This should be
```python
store, cleanup = _location_pull_files(plan.files, root)
try:
  ...
finally:
  cleanup()
```

in order to make this robust to exceptions. Or make `_location_pull_files` a context manager.

File: tests/helpers/__init__.py [hunk 15]
```diff
@@ -41,15 +48,109 @@ def create_random_packet(root, name="data", *, parameters=None, packet_id=None):
...
+            if i == 0 and base is None:
+                with open(packet_path / "data.txt", "w") as f:
+                    f.write("0")
+            else:
+                lines = [
+                    "with open('input.txt', 'r') as f:",
+                    "    data = f.read()",
+                    "with open('data.txt', 'w') as f:",
+                    f"    f.write(data + '{i}')",
+                ]
```
Comment (plietar) [inline tests/helpers/__init__.py]: Writing a script and evaluating it seems a bit overkill. Why don't we just write the data directly from this function?

File: tests/helpers/__init__.py [hunk 16]
```diff
@@ -41,15 +48,109 @@ def create_random_packet(root, name="data", *, parameters=None, packet_id=None):
     return p.id
 
 
+## Create a chain of packets a, b, c, ... that depend on each other
+def create_random_packet_chain(root, length, base=None):
```
Comment (plietar) [inline tests/helpers/__init__.py:50]: This could be a bit more concise and less noisy using the `create_packet` function above. You can also simplify the `id_use` code by just using the previous iteration's ID. Also just let the ID be chosen randomly by the `Packet` class.

```python
for i in range(length):
  with create_packet(root, name) as p:
     # Write some data
     if base is not None:
       packet.use_dependency(base, {"input.txt": "data.txt"})
  ids[name] = p.id
  base = p.id
return ids
```

File: src/outpack/location.py (hunk 3 above)
Comment (r-ash) [reply src/outpack/location.py]: Made that note a bit more detailed and included this in ticket for adding type annotations

File: src/outpack/index.py (hunk 2 above)
Comment (r-ash) [reply src/outpack/index.py:77]: Thanks added a ticket for that mrc-5043

File: src/outpack/filestore.py (hunk 1 above)
Comment (r-ash) [reply src/outpack/filestore.py:16]: Made a ticket for that, 5044

File: src/outpack/location_pull.py (hunk 6 above)
Comment (r-ash) [reply src/outpack/location_pull.py:198]: Thanks, this should be being used but is an oversight on my end! Have updated and fixed that now, also added an extra test for this.

File: src/outpack/location_pull.py (hunk 9 above)
Comment (r-ash) [reply src/outpack/location_pull.py]: Much nicer, thanks

File: src/outpack/location_pull.py (hunk 8 above)
Comment (r-ash) [reply src/outpack/location_pull.py:257]: Added a link to this into [mrc-4968](https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-4968)

File: src/outpack/location_pull.py (hunk 10 above)
Comment (r-ash) [reply src/outpack/location_pull.py]: Thanks, updated this in a couple of places and where I was using `-` for difference

File: tests/helpers/__init__.py (hunk 16 above)
Comment (r-ash) [reply tests/helpers/__init__.py:50]: Thanks, that's way better! I've used this and just written the data directly instead of writing a script as suggested above. Looking way better now

File: src/outpack/location_pull.py (hunk 12 above)
Comment (r-ash) [reply src/outpack/location_pull.py]: So I tried this initially but then got a lint issue "PEP 8: E731 do not assign a lambda expression, use a def". But I've dropped this now in favour of a context manager and setting a flag whether to cleanup the store (in the case that we're using a temporary filestore) or not.

File: src/outpack/filestore.py (hunk 13 above)
Comment (r-ash) [reply src/outpack/filestore.py]: Oof thanks for that, I've returned the `NamedTemporaryFile` as makes sense to use the already existing context manager for this.

**Comment by r-ash** [comment]
Thanks both for the excellent comments, think I've addressed those now and this is ready for another look.

File: src/outpack/filestore.py (hunk 13 above)
Comment (r-ash) [reply src/outpack/filestore.py]: Ok wow after a bit of a fight worked that out. Got to hand roll it yourself and manage the deletion manually as otherwise the file is deleted on windows before the context manager has exited.

Comment (plietar) [reply src/outpack/filestore.py]: I'm not sure I understand this. What do you mean by `otherwise the file is deleted on windows before the context manager has exited.`?

File: src/outpack/location_pull.py [hunk 17]
```diff
@@ -0,0 +1,465 @@
...
+            _location_pull_hash_store(
+                from_this_location,
+                location,
+                _location_driver(location, root),
+                store,
+            )
+
+    yield store
+    if cleanup_store:
+        store.destroy()
```
Comment (plietar) [inline src/outpack/location_pull.py]:
```suggestion
    try:
        yield store
    finally:
        if cleanup_store:
            store.destroy()
```

in case the yield (ie. the contents of the `with`-block) raises an exception.

File: src/outpack/search.py [hunk 18]
```diff
@@ -70,6 +71,8 @@ def as_query(query: Union[Query, str]) -> Query:
     if isinstance(query, Query):
         return query
     else:
+        if is_outpack_id(query):
+            query = f"'{query}'"
```
Comment (plietar) [inline src/outpack/search.py:75]: Is this something we want to add natively to the outpack parser?

**Review by plietar — APPROVED** [review]

File: src/outpack/search.py (hunk 18 above)
Comment (r-ash) [reply src/outpack/search.py:75]: Yeah I think we should do, I've made a ticket https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-5103/Search-for-packet-with-ID-without-quotes

File: src/outpack/filestore.py (hunk 13 above)
Comment (r-ash) [reply src/outpack/filestore.py]: Not best explanation - [from v 3.10 docs of NamedTemporaryFile ](https://docs.python.org/3.10/library/tempfile.html)

> That name can be retrieved from the name attribute of the returned file-like object. Whether the name can be used to open the file a second time, while the named temporary file is still open, varies across platforms (it can be so used on Unix; it cannot on Windows)

So on windows, we can't with v <3.12 create a named temporary file and then return the name, and open it later. As windows will have already deleted it. 

In [v3.12](https://docs.python.org/3/library/tempfile.html) they introduce a new arg `delete_on_close` which if `False` will only delete the file when the context manager exists (which is the behaviour we want here). But because we're running this on CI with older python versions we need to manage this deletion manually ourselves.

Comment (r-ash) [reply src/outpack/filestore.py]: I've fixed up the comment and linked to this discussion from the code for more context in future

**Comment by r-ash** [comment]
Ok quite a few changes a few new tests here to address coverage holes. Think ready for another look though

**Review by richfitz — APPROVED** [review]
